File: agentos-linux-wasm/wasm/x11-apps/netsurf_agentos_fetch.c
/*
 * Thin AgentOS network adapter for ORIGINAL NetSurf.
 *
 * Replaces curl's HTTP(S) registration (via --wrap=fetch_curl_register) with
 * fetches through the in-tab AgentOS proxy (/__agentos/proxy). Browser handles
 * TLS. Does not modify NetSurf browser core sources.
 */
#include <stdarg.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

#include "utils/nsurl.h"
#include "utils/corestrings.h"
#include "utils/errors.h"
#include "utils/ring.h"

#include "content/fetch.h"
#include "content/fetchers.h"

/* From NetSurf curl fetcher API */
nserror __real_fetch_curl_register(void);

extern int agentos_js_http_fetch(const char *url, const char *method,
				 const char *body, int body_len,
				 int *out_status, char **out_final_url,
				 uint8_t **out_body, int *out_body_len);

struct agentos_fetch_ctx {
	struct agentos_fetch_ctx *r_next, *r_prev;
	struct fetch *fetchh;
	nsurl *url;
	char *post_urlenc;
	bool aborted;
	bool locked;
	bool started;
};

static struct agentos_fetch_ctx *ring;

static bool agentos_initialise(lwc_string *scheme)
{
	(void)scheme;
	return true;
}

static void agentos_finalise(lwc_string *scheme)
{
	(void)scheme;
}

static bool agentos_can_fetch(const nsurl *url)
{
	(void)url;
	return true;
}

static void *agentos_setup(struct fetch *fetchh, nsurl *url, bool only_2xx,
			   bool downgrade_tls, const char *post_urlenc,
			   const struct fetch_multipart_data *post_multipart,
			   const char **headers)
{
	struct agentos_fetch_ctx *ctx;

	(void)only_2xx;
	(void)downgrade_tls;
	(void)post_multipart;
	(void)headers;

	ctx = calloc(1, sizeof(*ctx));
	if (!ctx)
		return NULL;
	ctx->fetchh = fetchh;
	ctx->url = nsurl_ref(url);
	if (post_urlenc)
		ctx->post_urlenc = strdup(post_urlenc);
	RING_INSERT(ring, ctx);
	return ctx;
}

static bool agentos_start(void *handle)
{
	struct agentos_fetch_ctx *ctx = handle;
	ctx->started = true;
	return true;
}

static void agentos_abort(void *handle)
{
	struct agentos_fetch_ctx *ctx = handle;
	ctx->aborted = true;
}

static void agentos_free(void *handle)
{
	struct agentos_fetch_ctx *ctx = handle;
	RING_REMOVE(ring, ctx);
	nsurl_unref(ctx->url);
	free(ctx->post_urlenc);
	free(ctx);
}

static void agentos_send_header(struct agentos_fetch_ctx *ctx, const char *fmt, ...)
{
	char buf[4096];
	va_list ap;
	fetch_msg msg;

	va_start(ap, fmt);
	vsnprintf(buf, sizeof(buf), fmt, ap);
	va_end(ap);

	msg.type = FETCH_HEADER;
	msg.data.header_or_data.buf = (const uint8_t *)buf;
	msg.data.header_or_data.len = strlen(buf);
	fetch_send_callback(&msg, ctx->fetchh);
}

static void agentos_process(struct agentos_fetch_ctx *ctx)
{
	fetch_msg msg;
	const char *url;
	const char *method = "GET";
	const char *body = NULL;
	int body_len = 0;
	int status = 0;
	char *final_url = NULL;
	uint8_t *resp = NULL;
	int resp_len = 0;
	int rc;

	url = nsurl_access(ctx->url);
	if (ctx->post_urlenc) {
		method = "POST";
		body = ctx->post_urlenc;
		body_len = (int)strlen(body);
	}

	ctx->locked = true;
	rc = agentos_js_http_fetch(url, method, body, body_len, &status,
				   &final_url, &resp, &resp_len);
	ctx->locked = false;

	if (ctx->aborted)
		goto done;

	if (rc != 0 || status <= 0) {
		msg.type = FETCH_ERROR;
		msg.data.error = final_url ? final_url : "AgentOS proxy fetch failed";
		fetch_send_callback(&msg, ctx->fetchh);
		goto done;
	}

	/* Proxy already followed redirects and returned the final body. */
	{
		const char *ctype = "text/html; charset=utf-8";
		if (final_url) {
			char *sep = strchr(final_url, '\0');
			/* meta is "url\0content-type" allocated as one buffer */
			if (sep && sep[1] != '\0')
				ctype = sep + 1;
		}

		fetch_set_http_code(ctx->fetchh, (http_response_code)status);

		if (!ctx->aborted)
			agentos_send_header(ctx, "HTTP/1.0 %d OK", status);
		if (!ctx->aborted)
			agentos_send_header(ctx, "Content-Type: %s", ctype);
		if (!ctx->aborted)
			agentos_send_header(ctx, "Content-Length: %d", resp_len);
	}

	if (!ctx->aborted && resp && resp_len > 0) {
		msg.type = FETCH_DATA;
		msg.data.header_or_data.buf = resp;
		msg.data.header_or_data.len = (size_t)resp_len;
		fetch_send_callback(&msg, ctx->fetchh);
	}

	if (!ctx->aborted) {
		msg.type = FETCH_FINISHED;
		fetch_send_callback(&msg, ctx->fetchh);
	}

done:
	free(final_url);
	free(resp);
}

static void agentos_poll(lwc_string *scheme)
{
	struct agentos_fetch_ctx *ctx, *save_ring = NULL;

	(void)scheme;
	while (ring != NULL) {
		ctx = ring;
		RING_REMOVE(ring, ctx);

		if (ctx->locked) {
			RING_INSERT(save_ring, ctx);
			continue;
		}
		if (!ctx->aborted && ctx->started)
			agentos_process(ctx);

		fetch_remove_from_queues(ctx->fetchh);
		fetch_free(ctx->fetchh);
	}
	ring = save_ring;
}

/* Linker --wrap replaces NetSurf's curl HTTP(S) registration. */
nserror __wrap_fetch_curl_register(void)
{
	const struct fetcher_operation_table ops = {
		.initialise = agentos_initialise,
		.acceptable = agentos_can_fetch,
		.setup = agentos_setup,
		.start = agentos_start,
		.abort = agentos_abort,
		.free = agentos_free,
		.poll = agentos_poll,
		.finalise = agentos_finalise,
	};
	nserror err;

	/* Do not call __real_fetch_curl_register — wasm curl has no TLS/sockets.
	 * AgentOS proxy (browser TLS) is the HTTP(S) path instead.
	 */
	(void)__real_fetch_curl_register;

	err = fetcher_add(lwc_string_ref(corestring_lwc_http), &ops);
	if (err != NSERROR_OK)
		return err;
	err = fetcher_add(lwc_string_ref(corestring_lwc_https), &ops);
	return err;
}

File: agentos-linux-wasm/wasm/x11-apps/netsurf_agentos_fetch.js
/**
 * Emscripten JS library: HTTP(S) fetch for original NetSurf via AgentOS proxy.
 * Browser does TLS; NetSurf browser core is unchanged.
 */
mergeInto(LibraryManager.library, {
  agentos_js_http_fetch__deps: ["$Asyncify", "malloc", "free"],
  agentos_js_http_fetch: function (
    urlPtr,
    methodPtr,
    bodyPtr,
    bodyLen,
    outStatusPtr,
    outFinalUrlPtrPtr,
    outBodyPtrPtr,
    outBodyLenPtr,
  ) {
    return Asyncify.handleAsync(async () => {
      const writeI32 = (ptr, value) => {
        HEAP32[ptr >> 2] = value | 0;
      };
      try {
        const url = UTF8ToString(urlPtr);
        const method = UTF8ToString(methodPtr) || "GET";
        let body = null;
        if (bodyPtr && bodyLen > 0) {
          body = UTF8ToString(bodyPtr, bodyLen);
        }
        const response = await fetch("/__agentos/proxy", {
          method: "POST",
          headers: { "content-type": "application/json" },
          body: JSON.stringify({
            type: "request",
            id: `netsurf-${Date.now()}-${Math.random().toString(16).slice(2)}`,
            url,
            method,
            headers: {
              "User-Agent":
                "Mozilla/5.0 (X11; Linux x86_64) NetSurf/3.11 AgentOS",
              Accept: "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
              "Accept-Language": "en-US,en;q=0.9",
              ...(method === "POST"
                ? { "Content-Type": "application/x-www-form-urlencoded" }
                : {}),
            },
            body,
          }),
        });
        if (!response.ok) {
          let message = `proxy HTTP ${response.status}`;
          try {
            const payload = JSON.parse(await response.text());
            if (payload && payload.error) message = payload.error;
          } catch (_) {
            /* keep status */
          }
          throw new Error(message);
        }
        const status = Number(response.headers.get("x-agentos-status") || "0");
        const finalUrl = response.headers.get("x-agentos-url") || url;
        let contentType = "text/html; charset=utf-8";
        try {
          const hdrJson = response.headers.get("x-agentos-response-headers");
          if (hdrJson) {
            const hdrs = JSON.parse(hdrJson);
            if (hdrs && hdrs["content-type"]) contentType = hdrs["content-type"];
            else if (hdrs && hdrs["Content-Type"]) contentType = hdrs["Content-Type"];
          }
        } catch (_) {
          /* keep default */
        }
        const bytes = new Uint8Array(await response.arrayBuffer());
        const bodyMem = _malloc(bytes.length + 1);
        if (!bodyMem) throw new Error("oom body");
        HEAPU8.set(bytes, bodyMem);
        HEAPU8[bodyMem + bytes.length] = 0;

        /* Pack final URL + content-type as "url\0content-type" for C side. */
        const meta = `${finalUrl}\0${contentType}`;
        const metaBytes = lengthBytesUTF8(meta) + 1;
        const metaMem = _malloc(metaBytes);
        if (!metaMem) {
          _free(bodyMem);
          throw new Error("oom meta");
        }
        stringToUTF8(meta, metaMem, metaBytes);

        writeI32(outStatusPtr, status);
        writeI32(outFinalUrlPtrPtr, metaMem);
        writeI32(outBodyPtrPtr, bodyMem);
        writeI32(outBodyLenPtr, bytes.length);
        return 0;
      } catch (err) {
        const msg = err && err.message ? String(err.message) : "fetch failed";
        const metaBytes = lengthBytesUTF8(msg) + 1;
        const metaMem = _malloc(metaBytes);
        if (metaMem) stringToUTF8(msg, metaMem, metaBytes);
        writeI32(outStatusPtr, 0);
        writeI32(outFinalUrlPtrPtr, metaMem || 0);
        writeI32(outBodyPtrPtr, 0);
        writeI32(outBodyLenPtr, 0);
        return -1;
      }
    });
  },
});
